modules/builders/src/prerender/worker.ts

```diff
@@ -0,0 +1,114 @@
+
+export function setup(options: WorkerSetupArgs): void {
+  workerArgs = options;
+  if (workerArgs.inlineCriticalCss) {
+    inlineCriticalCssProcessor = new InlineCriticalCssProcessor({
+      deployUrl: workerArgs.deployUrl,
+      minify: workerArgs.minifyCss,
+    });
+  }
+}
+
+/**
+ * Renders each route in routes and writes them to <outputPath>/<route>/index.html.
+ */
+export async function render(
+  outputPath: string,
+  serverBundlePath: string,
+  route: string,
+): Promise<void> {
+  const browserIndexOutputPath = path.join(outputPath, workerArgs.indexFile);
+  const outputFolderPath = path.join(outputPath, route);
+  const outputIndexPath = path.join(outputFolderPath, 'index.html');
+
+  const { renderModuleFn, AppServerModuleDef } = await getServerBundle(serverBundlePath);
+
+  const browserIndexInputPath = path.join(outputPath, workerArgs.indexFile);
+  let indexHtml = await fs.promises.readFile(browserIndexInputPath, 'utf8');
+  indexHtml = indexHtml.replace(
+    '</html>',
+    '<!-- This page was prerendered with Angular Universal -->\n</html>',
+  );
+  if (workerArgs.inlineCriticalCss) {
+    // Workaround for https://github.com/GoogleChromeLabs/critters/issues/64
+    indexHtml = indexHtml.replace(
+      / media=\"print\" onload=\"this\.media='all'"><noscript><link .+?><\/noscript>/g,
+      '>',
+    );
+  }
+
+  let html = await renderModuleFn(AppServerModuleDef, {
+    document: indexHtml,
+    url: route,
+  });
+
+  if (inlineCriticalCssProcessor) {
+    const { content, warnings, errors } = await inlineCriticalCssProcessor.process(html, {
+      outputPath,
+    });
+
+    // tslint:disable-next-line: no-non-null-assertion
+    warnings?.forEach((message) => process.send!({ logLevel: 'warn', message }));
+    // tslint:disable-next-line: no-non-null-assertion
+    errors?.forEach((message) => process.send!({ logLevel: 'error', message }));
+    html = content;
+  }
+
+  // This case happens when we are prerendering "/".
+  if (browserIndexOutputPath === outputIndexPath) {
+    const browserIndexOutputPathOriginal = path.join(outputPath, 'index.original.html');
+    fs.renameSync(browserIndexOutputPath, browserIndexOutputPathOriginal);
+  }
+
+  fs.mkdirSync(outputFolderPath, { recursive: true });
+  fs.writeFileSync(outputIndexPath, html);
+}
+
+/**
+ * Handles importing the server bundle.
+ */
+async function getServerBundle(bundlePath: string) {
+  const { AppServerModule, AppServerModuleNgFactory, renderModule, renderModuleFactory } =
+    await import(bundlePath);
+
+  if (renderModuleFactory && AppServerModuleNgFactory) {
+    // Happens when in ViewEngine mode.
+    return {
+      renderModuleFn: renderModuleFactory,
+      AppServerModuleDef: AppServerModuleNgFactory,
+    };
+  }
+
+  if (renderModule && AppServerModule) {
+    // Happens when in Ivy mode.
+    return {
+      renderModuleFn: renderModule,
+      AppServerModuleDef: AppServerModule,
+    };
+  }
```

Comment:
We no longer need to support VE

```suggestion
    return {
      renderModuleFn: renderModule,
      AppServerModuleDef: AppServerModule,
    };
```